src/outputs.rs

```diff
@@ -242,34 +249,66 @@ where
 
         op_builder.build(move |init_caps| {
             let parts: BTreeMap<StateKey, StatefulSink> = BTreeMap::new();
-            // Which partitions were written to in this epoch. We only
-            // snapshot those.
+            // Which partitions were written to in this epoch. We only snapshot those.
             let awoken: BTreeSet<StateKey> = BTreeSet::new();
 
-            let mut routed_tmp = Vec::new();
-            // First `StateKey` is partition, second is data
-            // routing.
+            // First `StateKey` is partition, second is data routing.
             type PartToInBufferMap = BTreeMap<StateKey, Vec<(StateKey, TdPyAny)>>;
             let mut items_inbuf: BTreeMap<S::Timestamp, PartToInBufferMap> = BTreeMap::new();
             let mut loads_inbuf = InBuffer::new();
             let mut ncater = EagerNotificator::new(init_caps, (parts, awoken));
 
+            let mut last_activation = Instant::now();
+            let timeout = timeout.map(|t| {
+                t.to_std()
+                    .expect("output's timeout should be a positive timedelta")
+            });
+            let mut routed_tmp: BTreeMap<(WorkerIndex, StateKey), Vec<(StateKey, TdPyAny)>> =
+                BTreeMap::new();
+
             move |input_frontiers| {
                 tracing::debug_span!("operator", operator = op_name).in_scope(|| {
                     routed_input.for_each(|cap, incoming| {
                         let epoch = cap.time();
-                        assert!(routed_tmp.is_empty());
-                        incoming.swap(&mut routed_tmp);
-                        for (worker, (part, (key, value))) in routed_tmp.drain(..) {
-                            assert!(worker == this_worker);
-                            items_inbuf
-                                .entry(*epoch)
-                                .or_insert_with(BTreeMap::new)
-                                .entry(part)
-                                .or_insert_with(Vec::new)
-                                .push((key, value));
+                        for (worker, (part, item)) in incoming.take() {
+                            routed_tmp
+                                .entry((worker, part))
```

Comment:
ok. I really want a way to test behavior around epoch boundaries programmatically, since I seem to miss issues related to this a lot of times, but now that you point it out I see the problem, thanks